src/GafferOSL/ShadingEngine.cpp

```diff
@@ -257,6 +259,88 @@ bool convertValue( void *dst, TypeDesc dstType, const void *src, TypeDesc srcTyp
 	return false;
 }
 
+struct ReplaceFunctor
+{
+	ReplaceFunctor( const IECore::CompoundObject *attributes ) : m_attributes( attributes )
+	{
+	}
+
+	std::string operator()( const boost::smatch & match )
+	{
+		// Search for attribute matching token
+		const StringData *sourceAttribute = m_attributes->member<StringData>( match[1].str() );
+		if( sourceAttribute )
+		{
+			return sourceAttribute->readable();
+		}
+		else
+		{
+			// Otherwise, return empty string
+			return "";
+		}
+	}
+
+	const IECore::CompoundObject *m_attributes;
+};
+
+boost::regex attributeRegex()
+{
+	// Extract ATTR_NAME from the pattern <attr:ATTR_NAME>
+	// Only match if the angle brackets haven't been escaped with a backslash
+	static boost::regex r( "(?<!\\\\)<attr:([^>]*[^\\\\>])>" );
+	return r;
+}
+
+std::string stringApplySubstitutions( const std::string &target, const IECore::CompoundObject *attributes )
+{
+	ReplaceFunctor rf( attributes );
+	std::string result = boost::regex_replace(
+		target, attributeRegex(), rf, boost::match_default | boost::format_all
+	);
+	boost::replace_all( result, "\\<", "<" );
+	boost::replace_all( result, "\\>", ">" );
+	return result;
+}
+
+void convertStringSubstitutions( ShaderNetwork *network )
+{
+	// Output parameters
+
+	std::map<InternedString, std::vector< ShaderNetwork::Parameter > > needSubstitution;
```

Comment:
Do we gain anything by using `InternedString` rather than `string` as the index here?

Reply:
Fixed in 0e361ab